### ISSUE TITLE
Implement a WebSocket RESTful API 

### ISSUE THREAD

**Issue by Nathanrs97** [sufst/intermediate-server#36]
### 1. Please describe the feature you are trying to request.
To optimise and improve the front-end GUI interface with the server, a RESTful API on top of a WebSocket would be a better-suited (and more standard) interface method. More background information can be seen in the front-end GUI issue [here](https://github.com/sufst/wireless-telemetry-gui/issues/3).

### 2. Describe the solution you'd like.
- A RESTful engine should be implemented that sits on top of a WebSocket and responds to API requests received over the WebSocket.
- The recently implemented SQLite staging database should be used as the interface between the RESTful engine and the Server protocol engine.
- The initial API only consists of the `GET \sensors?timesince=<epoch>&amount=<n>` where `timesince=<epoch>` determines when to select sensor data from, and `amount=<n>` determines the maximum number of sensor data points to respond with.
- The `GET` request simply pulls the wanted sensor data from the staging database.
- The format of the response to the `GET` request is 
```
{
	<sensor_name>:
	[
		{
		epoch: <float>,
		value: <int> or <bool>
		},
		...
	],
	...
}
```

**Comment by Nathanrs97** [comment]
It will be worth implementing the `PUT` API so sensor data can be loaded into the server for staging for `GET` requests. 

- Since bandwidth is limited on the XBee link it is unlikely that data-link will utilise this.
- The primary usage I see for the `PUT` API is either if the car is connected over ethernet (or any other high-speed IP-based network) or the front-end GUI wants to send data to the car (wireless isn't allowed but if we connect over ethernet then this is fine).

**Comment by AndreasDemenagas** [comment]
Another way you might go about implementing this, is having a new "event" sent by the WebSocket everytime a new PDU is available. 

For example maybe the server could send a `ACPDU` event every 0.5 seconds (or another set amount) and then the front-end client could then catch it it something like: 

```Javascript
socket.on('ACPDU', data => {
      ...
 });
```

So basically make each PDU have its own `.on(<event-name>)`

**Comment by Nathanrs97** [comment]
@AndreasDemenagas Unfortunately the RESTful is designed based on the client making requests to the server and the server responding, it doesn't allow for the server sending data to the client unprovoked. 

This is because RESTful is a state-less design. The server does not keep track of any states (like if a front-end client is connected to send periodic data to, for example).

I think it is best we keep to the RESTful design specification rather than modifying it, as then it becomes a question of where do we stop and when is it no longer a RESTful API. Unless you see any major problem which requires us to break the RESTful specification of the front-end making `GET` and `PUT` requests, we should stick to the specification.

**Comment by AndreasDemenagas** [comment]
Oh I see. Yes let's continue with the RESTful approach - this was just a random idea that came in my mind and thought might as well mention 😂

**Comment by Nathanrs97** [comment]
> Oh I see. Yes let's continue with the RESTful approach - this was just a random idea that came in my mind and thought might as well mention 😂

It was a good question, don't stop mentioning your ideas haha.

Just to expand a bit further. With the RESTful API, the front-end no longer needs to worry about PDUs as it just makes requests for `datasets` e.g. `/sensor` or `/sensor/RPM`.  This makes sense to me as PDUs are a more low-level protocol so it doesn't make a lot of sense the higher-level communication between the front-end and the server needs to worry about these.

**Comment by AndreasDemenagas** [comment]
So just to make sure I understand correctly (although I will surely have more questions in the future)

* Now that we will be moving to a WebSocket for the communication, in order to receive the data we will need to connect to a WebSocket URL - which I guess like be something like `ws://localhost:19900`? 
* Then we need to throw a request to the WebSocket with something like `GET /sensors` or `GET /sensors/RPM` depending on the values we need? 
* Also, I assume we will still need to send an AIPDU upon initial connection - right?

**Comment by Nathanrs97** [comment]
> - Now that we will be moving to a WebSocket for the communication, in order to receive the data we will need to connect to a WebSocket URL - which I guess like be something like ws://localhost:19900?
> - Then we need to throw a request to the WebSocket with something like GET /sensors or GET /sensors/RPM depending on the values we need?

Here is my python testing code for using the API which should clear things up.
```
async with websockets.connect("ws://localhost:8765") as websocket:
    await websocket.send("GET /sensors?amount=10")

    response_json = await websocket.recv()
    response = json.loads(response_json)

    print(response)
```
The response here returned will be the JSON of sensors and their data with 10 points in each (as we discussed previously).

> Also, I assume we will still need to send an AIPDU upon initial connection - right?

There is no need now because RESTful is state-less. The server has no use for keeping track of the front-end detail that was previously supplied with the AIDPU.

What we can do if you want to store front-end client details on the server (so like having a user name that corresponds to a person and their personal settings) is to use the `PUT` API e.g. `PUT /client/dave/show_rpm/False` which would set like a flag whether or not to show RPM or not (then you can access all settings with `GET /client/dave?` at start up).

**Comment by AndreasDemenagas** [comment]
Yes I took a look and it looks straight forward. It should not be hard at all to implement to the front-end. 

The idea about the `PUT`  request seems really nice for a potential further improvement after all the initial work is done.

**Comment by Nathanrs97** [comment]
> Yes I took a look and it looks straight forward. It should not be hard at all to implement to the front-end.

Not sure how long it is going to take you to update the front-end, but I plan to get a working `GET /sensor?amount=<n>` working this weekend so you should have that at least to test the front-end with.

**Comment by AndreasDemenagas** [comment]
So, I have successfully been able to connect to `ws://localhost:8765` and it looks like all the data is returned perfectly when I send `"GET /sensors?amount=10"` 🎉

So (just to make sure we are on the same page), we connect to it from the front-end and we send `"GET /sensors?amount=10"` to get the latest data. This that correct?

**Comment by Nathanrs97** [comment]
> So (just to make sure we are on the same page), we connect to it from the front-end and we send "GET /sensors?amount=10" to get the latest data. This that correct?

Yeah.

I will implement the filter `timesince=<epoch>` as I think will be easier for the front-end (lemme know if I'm wrong) because you can then just give the `epoch` time when you last did a `GET` so you'll only receive any new sensor data that has arrived since then. As `amount=<n>` will return sensor data the front-end would have received before if `n` is greater than the amount of sensor data arrived since the last `GET` request. It will also mean when you load the front-end you can just request the last 5 seconds worth of data or something.

**Comment by AndreasDemenagas** [comment]
Hmm surely the way to go would be using the `timesince=epoch`, however that means I would have to check for the latest epoch every time I receive data. 

Is there any other way of finding the latest epoch value rather than looping through all the JSON array returned and finding the max epoch? 

Or is my thinking completely wrong here?

**Comment by Nathanrs97** [comment]
According to [here](https://futurestud.io/tutorials/get-number-of-seconds-since-epoch-in-javascript) you can just do `const secondsSinceEpoch = Math.round(Date.now() / 1000)` to get the current epoch time in javascript. 

> Is there any other way of finding the latest epoch value rather than looping through all the JSON array returned and finding the max epoch?

You can't relay on the data timestamps as they will not be the current epoch time since they would have been from the past.

**Comment by AndreasDemenagas** [comment]
> Yeah, every time you receive data you would have to record locally the current epoch time, then use that next time you request data (this doesn't seem difficult to do?)

So every JSON object sent from the REST API has the same epoch across all of its objects?

**Comment by Nathanrs97** [comment]
> So every JSON object sent from the REST API has the same epoch across all of its objects?

Hmm, not completely sure what you mean by objects. But the `"time"` (epoch) field in each of the sensors data entries corresponds to the time that sensor data was captured. So every sensor "time" field across all its entries may be different.

This is what I was imagining:

- front-end loads up and requests last 5 seconds of data with like `GET \sensors?timesince={ Math.round(Date.now() / 1000) - 5)` and then record the Math.round(Date.now() / 1000).
- After like 0.5 seconds or whatever the refresh rate of the front-end you would do a `GET` request with the `timesince` equal the recorded value of last received data, so like `GET \sensors?timesince={lastEpoch}`.

Does this sound useful or we can scrap this and just go with `amount`? My only concern with only using `amount` is that the database is consistent (it is saved to file and loaded on server boot) so you can receive sensor data from the last time the server was used as the database will still have those values stored.

If using only `amount` is the best way for the front-end, then I will implement it so the server makes a new database every time the server is used so there is no stale data left over from previous runs. My only concern with this is that I was thinking of a feature in the future where we could use the front-end to load previous runs for viewing or saving to file.

**Comment by AndreasDemenagas** [comment]
Okay yes that makes sense. I was just confused.

**Comment by Nathanrs97** [comment]
> @Nathanrs97 Could I also ask - if it is possible - to not remove the classic TCP socket implementation, just so we can have an ultra ultra safe way back in case something goes completely wrong

Good thing you mentioned this as I thinking of removing it as it isn't optimal. 

I will leave it in until the front-end is reliably running on the RESTful API. if there are problems running the front-end GUI with the RESTful API (any problem really) then do please let me know and I can take a look as I do want to remove the TCP implemented of the PDU protocol for v0.1.0. 

I'm actually looking into removing the PDU protocol altogether in favour of either a RESTful or MQTT protocol as these are more optimal and are real standards for communications (so I have more trust that these will be easier to maintain in the long term than the custom PDU implementation).

**Comment by AndreasDemenagas** [comment]
Okay yes that makes sense. I will get working on bringing the RESTful API in the front-end asap. 

Surely when we get it to work properly in the GUI you can safely remove it.
